src/Common/src/Interop/Unix/libssl/Interop.SafeSslHandle.cs

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+
+internal static partial class Interop
+{
+    internal static partial class libssl
+    {
+        internal sealed class SafeSslContextHandle : SafeHandle
+        {
+            public SafeSslContextHandle(IntPtr method)
+                : base(IntPtr.Zero, true)
+            {
+                handle = SSL_CTX_new(method);
+            }
+
+            public override bool IsInvalid
+            {
+                get { return handle == IntPtr.Zero; }
+            }
+
+            protected override bool ReleaseHandle()
+            {
+                SSL_CTX_free(handle);
+                return true;
+            }
+        }
+
+        internal sealed class SafeSslHandle : SafeHandle
+        {
+            private readonly SafeBioHandle _readBio;
+            private readonly SafeBioHandle _writeBio;
+            private bool _isServer;
+
+            public bool IsServer
+            {
+                get {return _isServer;}
+            }
+
+            public SafeBioHandle InputBio
+            {
+                get
+                {
+                    return _readBio;
+                }
+            }
+
+            public SafeBioHandle OutputBio
+            {
+                get
+                {
+                    return _writeBio;
+                }
+            }
+
+            public SafeSslHandle(SafeSslContextHandle context, bool isServer)
+                : base(IntPtr.Zero, true)
+            {
+                _isServer = isServer;
+
+                IntPtr memMethod = Interop.libcrypto.BIO_s_mem();
+                _readBio = Interop.libcrypto.BIO_new(memMethod);
+                _writeBio = Interop.libcrypto.BIO_new(memMethod);
+                
+                IntPtr tempHandle = SSL_new(context);
+                if (tempHandle == IntPtr.Zero)
+                {
+                    return;
+                }
+
+                // After SSL_set_bio, the BIO handles are owned by SSL pointer
+                // and are automatically freed by SSL_free. To prevent a double
+                // free, we need to keep the ref counts bumped up till SSL_free
+                bool gotReadRef = false, gotWriteRef = false;
+                _readBio.DangerousAddRef(ref gotReadRef);
+                _writeBio.DangerousAddRef(ref gotWriteRef);
+                SSL_set_bio(tempHandle, _readBio.DangerousGetHandle(), _writeBio.DangerousGetHandle());
```

Comment:
If we were to get a ThreadAbortException between the DangerousAddRefs and the setting of the handle field the BIOs would never get freed.

```C#
bool gotReadRef = false, gotWriteRef = false;

try
{
    _readBio.DangerousAddRef(ref gotReadRef);
    ...
    handle = tempHandle;
}
catch
{
    if (gotWriteRef)
    {
        _writeBio.DangerousRelease();
    }

    ...

    throw;
}
```

Though you'd have better tracking, I think, if you made this ctor be a static method, and make SSL_new return a `SafeSslHandle`.  Then there's not also the problem of a ThreadAbortException needing to handle the situation where `tempHandle` was assigned but it never made it to `handle`.